File: web/src/types/ticket.ts
export type TicketPriority = "LOW" | "MEDIUM" | "HIGH" | "URGENT";

export type TicketStatus = "OPEN" | "IN_PROGRESS" | "RESOLVED" | "CLOSED";

export interface UserSummary {
  username: string;
  email: string;
  firstName: string;
  lastName: string;
}

export interface Ticket {
  id: number;
  title: string;
  description: string | null;
  priority: TicketPriority;
  status: TicketStatus;
  createdBy: UserSummary;
  createdAt: string;
  updatedAt: string;
}

export interface CreateTicketRequest {
  title: string;
  description?: string;
  priority: TicketPriority;
}

export interface UpdateTicketRequest {
  title?: string;
  description?: string;
  priority?: TicketPriority;
}

export interface UpdateStatusRequest {
  status: TicketStatus;
}

File: service/src/main/java/com/sethy/service/ticket/entity/Ticket.java
package com.sethy.service.ticket.entity;

import com.sethy.service.ticket.model.TicketPriority;
import com.sethy.service.ticket.model.TicketStatus;
import com.sethy.service.comment.entity.Comment;
import jakarta.persistence.*;
import lombok.Data;

import com.fasterxml.jackson.annotation.JsonIgnore;
import java.util.ArrayList;
import java.util.List;

import java.time.LocalDateTime;

@Entity
@Table(name = "tickets")
@Data
public class Ticket {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TicketPriority priority;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TicketStatus status = TicketStatus.OPEN;

    @Column(nullable = false, updatable = false)
    private String createdBy;

    @Column(updatable = false)
    private String createdByEmail;

    @Column(updatable = false)
    private String createdByFirstName;

    @Column(updatable = false)
    private String createdByLastName;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @JsonIgnore
    @OneToMany(mappedBy = "ticket", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<Comment> comments = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        // Set default values for user fields if not provided
        if (createdByEmail == null) createdByEmail = "";
        if (createdByFirstName == null) createdByFirstName = "";
        if (createdByLastName == null) createdByLastName = "";
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
